sdk/src/azure/core/az_log.c

```diff
@@ -16,75 +16,51 @@
 
 #ifndef AZ_NO_LOGGING
 
-static az_log_classification const* volatile _az_log_classifications = NULL;
+// Only using volatile here, not for thread safety, but so that the compiler does not optimize what
+// it falsely thinks are stale reads.
 static az_log_message_fn volatile _az_log_message_callback = NULL;
+static az_log_should_write_fn volatile _az_log_should_write_callback = NULL;
 
-void az_log_set_classifications(az_log_classification const classifications[])
-{
-  _az_log_classifications = classifications;
-}
-
-void az_log_set_callback(az_log_message_fn az_log_message_callback)
+void az_log_set_message_callback(az_log_message_fn az_log_message_callback)
 {
   _az_log_message_callback = az_log_message_callback;
 }
 
-// _az_LOG_WRITE_engine is a function private to this .c file; it contains the code to handle
-// _az_LOG_SHOULD_WRITE & _az_LOG_WRITE.
-//
-// If log_it is false, then the function returns true or false indicating whether the message
-// should be logged (without actually logging it).
-//
-// If log_it is true, then the function logs the message (if it should) and returns true or
-// false indicating whether it was logged.
-static bool _az_log_write_engine(bool log_it, az_log_classification classification, az_span message)
+void az_log_set_filter_callback(az_log_should_write_fn az_log_should_write_callback)
 {
-  // Copy the volatile fields to local variables so that they don't change within this function
-  az_log_message_fn const callback = _az_log_message_callback;
-  az_log_classification const* classifications = _az_log_classifications;
-
-  if (callback == NULL)
-  {
-    // If no one is listening, don't attempt to log.
-    return false;
-  }
-
-  if (classifications == NULL)
-  {
-    // If the user hasn't registered any classifications, then we log everything.
-    classifications
-        = &classification; // We don't need AZ_LOG_END_OF_LIST to be there, as very first comparison
-                           // is going to succeed and return from the function.
-  }
-
-  for (az_log_classification const* cls = classifications; *cls != AZ_LOG_END_OF_LIST; ++cls)
-  {
-    // If this message's classification is in the customer-provided list, we should log it.
-    if (*cls == classification)
-    {
-      if (log_it)
-      {
-        callback(classification, message);
-      }
-
-      return true;
-    }
-  }
-
-  // This message's classification is not in the customer-provided list; we should not log it.
-  return false;
+  _az_log_should_write_callback = az_log_should_write_callback;
 }
 
 // This function returns whether or not the passed-in message should be logged.
 bool _az_log_should_write(az_log_classification classification)
 {
-  return _az_log_write_engine(false, classification, AZ_SPAN_EMPTY);
+  _az_PRECONDITION(classification > 0);
```

Comment:
>  First, classification can be anything. Today it can't be <0, tomorrow it can, there's no special meaning behind that, no grand plan.

Exactly, once that is true _in the future_, we can remove the precondition. Weakening the precondition is all non-breaking. Also, this is an internal method that our code calls and guards against accidental bugs in our code, like a `Debug.Assert(...)`.

> Plus, this is a precondition in logging engine, which is already better to have non-faulty, because logging can be used during the error handling

I don't understand this. In some public APIs, whenever for simplicity, we removed az_result, we substituted the failure case with a precondition and docs making it the caller's responsibility to check and fix the bug in their code. The same holds true here.

> Plus, it is us who call these functions, so user does nothing wrong, it is our "bug", and has to deal with precondition failure, and what are they supposed to do?

Again, exactly. There is no harm in keeping it, and it helps us detects bugs in our code. This is conveying intent to other callers to make sure they set the log classification enum value correctly, and they aren't passing in a value that _currently_ doesn't make sense.

Reply:
I'll let you decide, but it is much harder to get an invalid classification, than to get an invalid span. But we will be checking this up to three times during each logging attempt (should_log, then log), and if we change the bit representation to allow more values, nothing tells us to update this place - there's nothing like _az_PRECONDITION_IS_VALID_CLASSIFICATION(), so the only way to find out would be to trip on these preconditions.